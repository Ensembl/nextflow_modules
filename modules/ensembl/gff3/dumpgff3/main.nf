// See the NOTICE file distributed with this work for additional information
// regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

process GFF3_DUMPGFF3 {
    tag "${db.species}"
    label 'process_low'
    // maxForks "${params.max_database_forks}"
    
    container 'ensemblorg/ensembl-legacy-scripts:e112_APIv0.4'

    input:
        val db

    output:
        tuple val(db), val("gff3"), path("*.gff3"), emit: gff3
        path "versions.yml", emit: versions

    when:
        task.ext.when == null || task.ext.when

    script:
        version = "0.4"
        output = "${db.species}.gff3"
        temp_gff = "temp.gff3"
        password_arg = db.server.password ? "--pass ${db.server.password}" : ""
        """
        dump_gff3.pl \
            --host ${db.server.host} \
            --port ${db.server.port} \
            --user ${db.server.user} \
            ${password_arg} \
            --dbname ${db.server.database} \
            > ${temp_gff}

        # Tidy up
        gt gff3 -tidy -sort -retainids ${temp_gff} > ${output}
        gt gff3validator ${output}
        rm ${temp_gff}

        echo -e -n "${task.process}:\n\tensembl-genomio: ${version}" > versions.yml
        """

    stub:
        version = "0.4"
        output_stub = "gff_outfile.gff3"
        input_gff3 = "${projectDir}/tests/module/ensembl/gff3/dumpgff3/test_dump.gff3"
        """
        cp ${input_gff3} ${output_stub}

        echo -e -n "${task.process}:\n\tensembl-genomio: ${version}" > versions.yml
        """
}
