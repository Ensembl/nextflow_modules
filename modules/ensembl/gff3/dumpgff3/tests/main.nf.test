// See the NOTICE file distributed with this work for additional information
// regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// nf-core modules test gff3/dumpgff3
nextflow_process {

    name "Test Process GFF3_DUMPGFF3"
    script "../main.nf"
    process "GFF3_DUMPGFF3"

    tag "modules"
    tag "modules_local"
    tag "gff3"
    tag "gff3/dumpgff3"

    test("GFF3 Dump gene models [stub]") {

        when {
            options "-stub"

                process {
                    """
                    input[0] = [[ test_id:'dummy_core_db_core_110_1', species:'Genus species']]
                    """
                }
            }

            then {
                assert snapshot(
                    process.out.gff3.collect {
                        it.collect { it instanceof String && file(it).exists() ? file(it).name : it }
                        }
                    ).match("gff3")

            assertAll(
                    { assert process.success },
                    { assert snapshot(process.out).match() },
                    { assert process.out.gff3.size() == 1 },
                    { assert process.out.gff3 != null }
            )
        }
    }
}
