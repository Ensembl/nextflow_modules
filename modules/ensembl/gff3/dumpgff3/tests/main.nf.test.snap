{
    "GFF3 Dump gene models [stub]": {
        "content": [
            {
                "0": [
                    [
                        [
                            {
                                "test_id": "dummy_core_db_core_110_1",
                                "species": "Genus species"
                            }
                        ],
                        "gff3",
                        "gff_outfile.gff3:md5,e8261f55873c877206959c86cc91c678"
                    ]
                ],
                "1": [
                    "versions.yml:md5,f781e978faca3da286270de33f6677a9"
                ],
                "gff3": [
                    [
                        [
                            {
                                "test_id": "dummy_core_db_core_110_1",
                                "species": "Genus species"
                            }
                        ],
                        "gff3",
                        "gff_outfile.gff3:md5,e8261f55873c877206959c86cc91c678"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f781e978faca3da286270de33f6677a9"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-12-16T16:14:41.80428652"
    },
    "gff3": {
        "content": [
            [
                [
                    [
                        {
                            "test_id": "dummy_core_db_core_110_1",
                            "species": "Genus species"
                        }
                    ],
                    "gff3",
                    "gff_outfile.gff3"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-12-16T16:14:41.678407952"
    }
}